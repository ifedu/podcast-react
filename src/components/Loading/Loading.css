.Loading img {
  width: 24px;
}
