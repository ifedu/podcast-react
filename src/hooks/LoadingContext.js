import { createContext, useContext, useState } from 'react';

const LoadingContext = createContext({
  isLoading: false,
  setIsLoading: null,
});

export function LoadingProvider({ children }) {
  const [isLoading, setIsLoading] = useState(false);

  return(<LoadingContext.Provider value={{ isLoading, setIsLoading }}>{children}</LoadingContext.Provider>)
}

export function useLoading() {
  return useContext(LoadingContext);
}
