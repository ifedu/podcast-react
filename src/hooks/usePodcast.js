import React from 'react';

export const usePodcast = (setIsLoading, podcastId) => {
  const [ podcastInfo, setPodcast ] = React.useState(null);

  React.useEffect(() => {
    setIsLoading(true);

    if (
      localStorage.getItem(`last podcast ${podcastId}`) !== null &&
      Math.floor((new Date() - new Date(localStorage.getItem(`last podcast ${podcastId}`))) / (1000 * 60 * 60 * 24)) === 0
    ) {
      setPodcast(JSON.parse(localStorage.getItem(`podcast ${podcastId}`)));
      setIsLoading(false);
      return;
    }

    fetch(`https://api.allorigins.win/get?url=${encodeURIComponent(`https://itunes.apple.com/lookup?id=${podcastId}&media=podcast&entity=podcastEpisode&limit=20`)}`)
    .then((resp) => resp.json())
    .then((resp) => {
      localStorage.setItem(`podcast ${podcastId}`, resp.contents);
      localStorage.setItem(`last podcast ${podcastId}`, new Date());

      setPodcast(JSON.parse(resp.contents));
      setIsLoading(false);
    });
  }, [setIsLoading, podcastId]);

  return podcastInfo;
}
