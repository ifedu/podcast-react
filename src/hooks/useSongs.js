import React from 'react';

export const useSongs = (setIsLoading) => {
  const [songs, setSongs] = React.useState(null);

  React.useEffect(() => {
    setIsLoading(true);

    if (
      localStorage.getItem('last songs') !== null &&
      Math.floor((new Date() - new Date(localStorage.getItem('last songs'))) / (1000 * 60 * 60 * 24)) === 0
    ) {
      setSongs(JSON.parse(localStorage.getItem('songs')));
      setIsLoading(false);
      return;
    }

    fetch('https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json')
    .then((resp) => resp.json())
    .then((resp) => {
      localStorage.setItem('songs', JSON.stringify(resp.feed));
      localStorage.setItem('last songs', new Date());

      setSongs(resp.feed);
      setIsLoading(false);
    });
  }, [setIsLoading]);

  return songs;
}
