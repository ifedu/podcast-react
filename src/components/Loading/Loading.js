import React from 'react';

import './Loading.css';
import loading from '../../assets/loading.gif';
import { useLoading } from '../../hooks/LoadingContext';

export function Loading() {
  const { isLoading } = useLoading();

  return (
    <div className='Loading'>
      {isLoading && <img src={loading}></img>}
    </div>
  );
}
