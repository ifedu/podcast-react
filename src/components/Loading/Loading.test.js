import { cleanup, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer' ;

import { Loading } from './Loading';
import { useLoading } from '../../hooks/LoadingContext';

jest.mock('../../hooks/LoadingContext.js');

describe.only('Loading', () => {
  const view = <Loading/>;

  afterEach(cleanup);

  it('renders correctly', () => {
    useLoading.mockImplementation(jest.fn(() => ({ isLoading: true })));

    const tree = renderer.create(view).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the loading img', async() => {
    useLoading.mockImplementation(jest.fn(() => ({ isLoading: true })));

    render(view);

    await waitFor(() => {
      const loading = screen.queryByTestId('loading-img');
      expect(loading).toBeInTheDocument();
    });
  });

  it("doesn't render the loading img", async() => {
    useLoading.mockImplementation(jest.fn(() => ({ isLoading: false })));

    render(view);

    await waitFor(() => {
      const loading = screen.queryByTestId('loading-img');
      expect(loading).toBeNull();
    });
  });
});
