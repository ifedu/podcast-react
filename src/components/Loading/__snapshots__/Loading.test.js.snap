// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading renders correctly 1`] = `
<div
  className="Loading"
>
  <img
    data-testid="loading-img"
    src="loading.gif"
  />
</div>
`;
